src/main/java/com/dumbdogdiner/stickyapi/bukkit/book/chat/ClickEvent.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2020-2021 DumbDogDiner <dumbdogdiner.com>. All rights reserved.
+ * Licensed under the MIT license, see LICENSE for more information...
+ */
+package com.dumbdogdiner.stickyapi.bukkit.book.chat;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.experimental.UtilityClass;
+
+/**
+ * An event that is fired when text associated with this event is clicked.
+ */
+public abstract class ClickEvent {
+    private ClickEvent() {} // seal this class
+
+    /**
+     * Convert the ClickEvent into JSON representation used in-game.
+     * @return {@link JsonElement}
+     */
+    public abstract @NonNull JsonElement toJson();
+
+    /**
+     * Create a copy of this ClickEvent.
+     * @return {@link ClickEvent}
+     */
+    @Override
+    public abstract @NonNull ClickEvent clone();
+
+    /**
+     * When clicked, displays a prompt to open a URL in the player's browser.
+     */
+    public final static class OpenUrl extends ClickEvent {
+        /** The URL to open. */
+        @Getter
+        private final @NonNull String url;
+
+        public OpenUrl(@NonNull String url) {
+            this.url = url;
+        }
+
+        @Override
+        public JsonElement toJson() {
+            JsonObject object = new JsonObject();
+            object.addProperty("action", "open_url");
+            object.addProperty("value", url);
+            return object;
+        }
+
+        @Override
+        public OpenUrl clone() {
+            return new OpenUrl(url);
+        }
+    }
+
+    /**
+     * When clicked, opens a file on the player's computer. This event can only be used by messages generated by the
+     * game, and will not function otherwise.
+     */
+    public final static class OpenFile extends ClickEvent {
```

Comment:
Security isn't my best topic, but this seems like it could present a massive security hole...